src/autodrop.hpp

```diff
@@ -0,0 +1,32 @@
+template<class T>
```

Comment:
Why are you creating your own smart pointer implementation? Just use something from the STL.